File: solutions/zig/perfect-numbers/1/perfect_numbers.zig
pub const Classification = enum {
    deficient,
    perfect,
    abundant,
};

fn aliquot_sum(n: u64) u64 {
    var sum: usize = 0;
    if (n > 1) for (1..n - 1) |f| {
        if (n % f == 0) sum += f;
    };
    return sum;
}

/// Asserts that `n` is nonzero.
pub fn classify(n: u64) Classification {
    const a = aliquot_sum(n);
    if (a < n) return .deficient;
    if (a > n) return .abundant;
    return .perfect;
}

File: solutions/lua/perfect-numbers/1/perfect-numbers.lua
local function aliquot_sum(n)
  local a = 0
  for f = 1, n / 2 do
    if n % f == 0 then a = a + f end
  end
  return a
end

local function classify(n)
  local a = aliquot_sum(n)
  if a < n then return "deficient" end
  if a > n then return "abundant" end
  return "perfect"
end

return { aliquot_sum = aliquot_sum, classify = classify }
